src/Compilers/CSharp/Test/Emit3/Semantics/ExtensionTests.cs

```diff
@@ -21304,6 +21306,1577 @@ static class E
+        public void M4<T, ts>(string s) { }           // Error: Cannot reuse names `T` and `ts`
+    }
+}
+";
+        var comp = CreateCompilation(src);
+
+        comp.VerifyDiagnostics(
+            // (9,39): error CS9512: Cannot use extension parameter 'T[] ts' in this context.
+            //         public static bool M2(T t) => ts.Contains(t); // Error: Cannot refer to `ts` from static context
+            Diagnostic(ErrorCode.ERR_InvalidExtensionParameterReference, "ts").WithArguments("T[] ts").WithLocation(9, 39),
+            // (10,28): error CS9507: 'T': a parameter, local variable, or local function cannot have the same name as an extension container type parameter
+            //         public void M3(int T, string ts) { }          // Error: Cannot reuse names `T` and `ts`
+            Diagnostic(ErrorCode.ERR_LocalSameNameAsExtensionTypeParameter, "T").WithArguments("T").WithLocation(10, 28),
+            // (10,38): error CS9509: 'ts': a parameter, local variable, or local function cannot have the same name as an extension parameter
+            //         public void M3(int T, string ts) { }          // Error: Cannot reuse names `T` and `ts`
+            Diagnostic(ErrorCode.ERR_LocalSameNameAsExtensionParameter, "ts").WithArguments("ts").WithLocation(10, 38),
+            // (11,24): error CS9508: Type parameter 'T' has the same name as an extension container type parameter
+            //         public void M4<T, ts>(string s) { }           // Error: Cannot reuse names `T` and `ts`
+            Diagnostic(ErrorCode.ERR_TypeParameterSameNameAsExtensionTypeParameter, "T").WithArguments("T").WithLocation(11, 24),
+            // (11,27): warning CS8981: The type name 'ts' only contains lower-cased ascii characters. Such names may become reserved for the language.
+            //         public void M4<T, ts>(string s) { }           // Error: Cannot reuse names `T` and `ts`
+            Diagnostic(ErrorCode.WRN_LowerCaseTypeName, "ts").WithArguments("ts").WithLocation(11, 27),
+            // (11,27): error CS9511: Type parameter 'ts' has the same name as an extension parameter
+            //         public void M4<T, ts>(string s) { }           // Error: Cannot reuse names `T` and `ts`
+            Diagnostic(ErrorCode.ERR_TypeParameterSameNameAsExtensionParameter, "ts").WithArguments("ts").WithLocation(11, 27)
+            );
+    }
+
+    [Fact]
+    public void NameConflict_27_ExampleFromSpec()
+    {
+        var src = @"
+public static class E
+{
+    extension<T>(T[] ts)
+    {
+        public int T() { return M(ts); } // Generated static method M<T>(T[]) is found
+        public string M() { return T(ts); } // Error: T is a type parameter
+    }
+}
+
+class CTest
+{
+    static int M<T>(T[] ts)
+    {
+        return T(ts);
+    }
+    
+    static int T<U>(U[] ts) => 0;
+}
+";
+        var comp = CreateCompilation(src);
+
+        comp.VerifyDiagnostics(
```

Comment:
>VerifyDiagnostics [](http://example.com/codeflow?start=13&length=17)

nit: consider `VerifyEmitDiagnostics` over `VerifyDiagnostics`. Applies to other added tests too